### ISSUE TITLE
openssl verify silently ignores any but the first certificate in the certificates argument

### ISSUE THREAD

**Issue by vanschelven** [openssl/openssl#14675]
Steps to reproduce:

```
$ cp /etc/ssl/certs/Certigna_Root_CA.pem /tmp/poc.pem
$ echo "unverifiable garbage" >> /tmp/poc.pem
$ openssl verify /tmp/poc.pem 
/tmp/poc.pem: OK
```

(In actuality "unverifiable garbage" would more likely be some other certificate, which could be either malformed or malicious)

In the manual the final argument to `verify` is called `certificates`, which is documented like so:

```
      certificates
           One or more certificates to verify. 
```

Presumably this is to be read as "if multiple arguments are provided, each will be individually checked as a leaf certificate". From the above it is clear that it is _not_ to be read as "a single file containing multiple certificates"... but IMHO it could easily be read as such.

This violates user expectations in way that has obvious security implications. Some evidence from the internet about the confusion:

* https://unix.stackexchange.com/a/229733/64671 (the answer suggests calling `openssl verify` on a bundle)
* https://stackoverflow.com/questions/20409534/how-does-an-ssl-certificate-chain-bundle-work (the question is about calling `openssl verify` on a bundle, and none of the answer to date points at the actual flaw) 

Version info:

```
$ openssl version
OpenSSL 1.1.1f  31 Mar 2020
```

**Comment by richsalz** [comment]
Yes, it is multiple files one each having one certificate.  In master this is a little better, but there it should say, somewhere "one per file" in both versions.

**Comment by vanschelven** [comment]
I've made a PR for the documentation that makes this slightly more clear. IMHO an explicit error is really the way to go here, but I'm not sufficiently confident in my understanding of the codebase to provide a PR for that myself.

**Comment by t8m** [comment]
I do not think we want to change the behavior. But yeah, documentation needs to be improved.

**Comment by vanschelven** [comment]
> I do not think we want to change the behavior.

I have to admit this is surprising to me... what's the rationale?

**Comment by vkopichenko** [comment]
> I do not think we want to change the behavior. But yeah, documentation needs to be improved.

Hi, @t8m.
The links provided in the issue illustrate that current behavior is not the most intuitive and even fixing the docs would not make it less complicated as could be.
Maybe you could consider providing a new handy way to verify the whole certificate chain in a single command. It may be done under some new switch for compatibility reasons.

**Comment by levitte** [comment]
Taking some inspiration from letsencrypt, a flag `-fullchain` could be added to flag that the input file is assumed to contain a full chain.

**Comment by t8m** [comment]
> > I do not think we want to change the behavior.
> 
> I have to admit this is surprising to me... what's the rationale?

This has a potential of breaking existing applications and scripts. But yeah, making the new behavior optional with some flag such a `-fullchain` would be fine. If someone submits a PR it certainly will be considered.

**Comment by kaduk** [comment]
I remember looking at the `x509` utility to let it look at more certs than just the first one in the file, and the structure of the existing code was going to make that a substantial effort.  I mostly assume that the `verify` code is going to be in a similar position.  Which is not to say that we wouldn't welcome patches, just that it may not be as trivial to implement as it sounds like it should.

**Comment by nhorman** [comment]
Marking as inactive, to be closed at the end of 3.4 dev barring further input

**Comment by richsalz** [comment]
Oh c'mon.  This is a one-line fix.  The project can't do it?

**Comment by nhorman** [comment]
Can we? Sure.  will it get scheduled in light of the rest of the issues we're addressing?  I don't think so.  We don't prioritize the easy stuff first, just because its easy.

That said:
1) My read of the above commentary made it seem like this issue was less obvious than you make it out to be.  I could be wrong of course, I'm trying to make it through all 2000 bugs we have.
2) I'm happy to mark this as a community issue, and would welcome a PR if you're willing to provide one.  Interested?

**Comment by richsalz** [comment]
At this point, the issue is purely documentation:  stuff is skipped until the first PEM marker is found, and stuff after the closing PEM marker is ignored.  That is important behavior to document and has been around since SSLeay days.  It lets you do `openssl x509 -text ... ` and feed the output into anything that wants a cert in PEM format, among other things.

I don't know the right place to document it.  Probably in `doc/man3/PEM_read.pod`. I hope the project will do so.

**Comment by nhorman** [comment]
Thats fair, and a clear statement of what should be done.  I still don't think we will likely get to this, as it should be addressable by any community member that is interested in getting started with openssl.  Perhaps we will make doc fixups a planning theme for a future minor release, unsure.  Until then, I'm marking this as a community issue.

**Comment by vanschelven** [comment]
1. _some_ (though perhaps not a complete) documentation fix was already provided and merged.
2. the above remark "It lets you do openssl x509 -text ... and feed the output into anything that wants a cert in PEM format, among other things." is, I think, the first time an actual argument was provided for the current behavior, however:
3. I still think "silently ignoring" is not a good default in general, and _especially not for a security-oriented library_.
4. evidence that the current behavior is highly confusing was already provided in the links above. To wit, in the 2 links provided above 28 and 33 people respectively found the wrong (as per the current issue) answer convincing enough to upvote it. In both cases the wrong answer is the top answer. The number of people not bothered to upvote and/or just reading the answer will likely be orders of magnitude higher.

Anywayssss, the maintainers of this project are free to treat issues as they see fit. But I'm also free to draw my own conclusions about the security-awareness of the project then, as well as inform others of the same :smile:

**Comment by t8m** [comment]
We could potentially do something here - we still need to keep ignoring _garbage_ outside the PEM markers but we could simply attempt to read another certificate from the file after the first one and if that succeeds we could print a warning that this and any following certificates are ignored. It would not be a bug fix but a feature that could go into a minor release.

**Comment by richsalz** [comment]
@vanschelven the original intent of the PEM encoding was that it be appended to mail messages; see https://www.rfc-editor.org/rfc/rfc1113.txt written in 1989.  You could feel "free to draw [your] own conclusions about the security-awareness of the project then, as well as inform others of the same"  But understand that you are complaining about the industry standard and people might think you are not very well informed.

**Comment by vanschelven** [comment]
> we still need to keep ignoring garbage outside the PEM markers but we could simply attempt to read another certificate from the file after the first one and if that succeeds we could print a warning that this and any following certificates are ignored. 

To me, this would seem like a solution that satisfies everyone.

**Comment by t8m** [comment]
Please note we are looking for community help implementing this issue. Please comment if you are interested in implementing this feature or if you have reasons why this feature is important enough so we should consider implementing it with higher priority than other requested features. The issue will be closed after 3 months otherwise.

**Comment by t8m** [comment]
This issue was closed because it was marked with the help wanted label and there was no interest in implementing this issue or any other activity for more than 2 months.
